source/_posts/2019-07-15-complete-flows-partial-models.markdown

```diff
@@ -0,0 +1,166 @@
+    return FormData(name: partialFormData.name, age: age, password: nil)
+}
+
+func thirdStepFinished(with password: String, partialFormData: FormData) {
+    let formData = FormData(name: partialFormData.name, age: partialFormData.age, password: password)
+    
+    api.performLogin(with: formData)
+}
+```
+
+However, now we need to `guard` against any `nil` values if we want to use them (for example, to make a network request).
+
+```swift
+guard let name = formData.name,
+      let age = formData.age,
+      let password = formData.password {
+    return // what should we do here???
+}
+
+// use data
+```
+
+From a domain perspective, that `return` doesn't make any sense.
+One may say, _"we can raise an error to the user"_ or _"we could track it and check if users are getting stuck somehow"_. But, at the end of the day, this is not a good solution because you know that when the flow ends, you have all the values.
+
+Our model is "lying" to us. That's not loosen, it's just flawed.
+
+There are several approaches to make it better, like "one model per step":
+
+```swift
+struct FirstStep {
+    let name: String
+}
+
+struct SecondStep {
+    let name: String
+    let age: Int
+    
+    init(firstStep: FirstStep, age: Int) {
+        self.name = firstStep.name
+        self.age = age
+    }
+}
+
+struct ThirdStep {
+    let name: String
+    let age: Int
+    let password: String
+    
+    init(secondStep: SecondStep, password: String) {
+        self.name = secondStep.name
+        self.age = secondStep.age
+        self.password = password
+    }
+}
+```
+
+That's better! But there is also another way of doing things that doesn't involve duplication nor partial data structs.
+
+Instead of breaking down our data structure, why not to break down functions?
+
+Our `FormData` initializer, when interpreted as a function, has this shape:
+
+```swift
+(String, Int, String) -> FormData
+```
+
+But we can break it down into plain old lambdas[^1], and by applying it to the initializer for our data structure:
+
+```swift
+(String) -> (Int) -> (String) -> FormData
+```
+
+This technique is called [currying](https://www.pointfree.co/episodes/ep5-higher-order-functions#t42). What it does is, it allow us to translate the evaluation of a function that takes multiple arguments into evaluating a sequence of functions, each with a single argument.
+
+```swift
+func curry<A, B, C, D>(
+    _ f: @escaping (A, B, C) -> D
+) -> (A) -> (B) -> (C) -> D {
+    return { a in { b in { c in return f(a, b, c) } } }
+}
+```
+
+The function above goes from a function that takes multiple arguments `(A, B, C)` and produces a `D`, to single functions, that take one argument each: `(A) -> (B) -> (C)` and produces a `D`, making it possible to partially apply each argument, one at the time, until it can evaluate and return the output value.
+
+Using it in our flow, may look like the following:
+
+```swift
+typealias FromFirstStep = (Int) -> (String) -> FormData
```

Comment:
How about typealiasing also `name`, `age` and `password` like this:
```swift
typealias Name = String
typealias Age = Int
typealias Password = String
```
After that, we could have something like:
```swift
typealias FromFirstStep = (Age) -> (Name) -> FormData
```
and so on :) way easier to read IMHO

Reply:
I can't express how much I love this.